### PR TITLE
[SPARK-27749][SQL] hadoop-3.2 support hive-thriftserver

### PR DESCRIPTION
**Pull request by wangyum**
## What changes were proposed in this pull request?

This PR mainly makes the following changes to make `hadoop-3.2` support `sql/hive-thriftserver`:
1. Upgrade [`TCLIService.thrift`](https://github.com/apache/hive/blob/rel/release-2.3.5/service-rpc/if/TCLIService.thrift) and related code to Hive 2.3.5 because of [HIVE-12442](https://issues.apache.org/jira/browse/HIVE-12442)(Note that we only migrate code without adding features, such as [HIVE-4924](https://issues.apache.org/jira/browse/HIVE-4924) and [HIVE-15473](https://issues.apache.org/jira/browse/HIVE-15473)).
2. Use slf4j as logging facade because of [HIVE-12237](https://issues.apache.org/jira/browse/HIVE-12237).
3. Port [HIVE-13169](https://issues.apache.org/jira/browse/HIVE-13169) to compatible with Hive 2.3.

## How was this patch tested?

Exiting test

### CODE REVIEW

**Comment by wangyum** [comment]
retest this please

**Comment by dongjoon-hyun** [comment]
@wangyum . I reopened https://issues.apache.org/jira/browse/SPARK-27402 . Please make an independent follow-up PR using ID `SPARK-27402`. That will give us traceability.

File: sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/CliSuite.scala [hunk 1]
```diff
@@ -200,10 +201,13 @@ class CliSuite extends SparkFunSuite with BeforeAndAfterAll with Logging {
   }
 
   test("Commands using SerDe provided in --jars") {
+    val jarName = if (HiveUtils.isHive23) {
+      "hive-hcatalog-core-2.3.4.jar"
```
Comment (dongjoon-hyun) [inline sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/CliSuite.scala]: `"hive-hcatalog-core-2.3.5.jar"`?

File: sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/HiveThriftServer2Suites.scala [hunk 2]
```diff
@@ -484,10 +483,13 @@ class HiveThriftBinaryServerSuite extends HiveThriftJdbcTest {
     withMultipleConnectionJdbcStatement("smallKV", "addJar")(
       {
         statement =>
+          val jarName = if (HiveUtils.isHive23) {
+            "hive-hcatalog-core-2.3.4.jar"
```
Comment (dongjoon-hyun) [inline sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/HiveThriftServer2Suites.scala]: ditto. `"hive-hcatalog-core-2.3.5.jar"`?

File: sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/HiveThriftServer2Suites.scala [hunk 3]
```diff
@@ -538,7 +540,11 @@ class HiveThriftBinaryServerSuite extends HiveThriftJdbcTest {
         conf += resultSet.getString(1) -> resultSet.getString(2)
       }
 
-      assert(conf.get("spark.sql.hive.version") === Some("1.2.1"))
+      if (HiveUtils.isHive23) {
+        assert(conf.get("spark.sql.hive.version") === Some("2.3.4"))
```
Comment (dongjoon-hyun) [inline sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/HiveThriftServer2Suites.scala]: `Some("2.3.5")`?

File: sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/HiveThriftServer2Suites.scala [hunk 4]
```diff
@@ -551,7 +557,11 @@ class HiveThriftBinaryServerSuite extends HiveThriftJdbcTest {
...
```
Comment (dongjoon-hyun) [inline sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/HiveThriftServer2Suites.scala]: `Some("2.3.5")`?

**Comment by dongjoon-hyun** [comment]
So, the files under `v2.3.5` are from Hive `2.3.5`, not Hive `2.3.4`, right?

File: sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/HiveThriftServer2Suites.scala (hunk 4 above)
Comment (dongjoon-hyun) [reply sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/HiveThriftServer2Suites.scala]: Yep. `2.3.4` causes UT failure.

**Review by dongjoon-hyun — CHANGES_REQUESTED** [review]
Please replace `2.3.4` to `2.3.5` to pass the UTs. Thanks.

cc @gatorsmile since this is a massive change.

File: sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/CliSuite.scala (hunk 1 above)
Comment (wangyum) [reply sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/CliSuite.scala]: Yes. Thank you @dongjoon-hyun

**Comment by dongjoon-hyun** [comment]
@wangyum . Just to be clear, did you regenerate Thrift related code always for Spark?
> (Note that we only migrate code without adding features, such as HIVE-4924 and HIVE-15473).

I assumed that this generate code comes from Hive 3.5.3, but the generated code is not the same with Hive 3.5.3.

cc @gatorsmile

**Comment by wangyum** [comment]
Yes @dongjoon-hyun I did [some changes](https://github.com/wangyum/spark/commit/638c47fabaaa1a07dad9a7a2a376e02cb7551046) to workaround `mvn install` build failure with error: `not found: type AsyncFrameBuffer...`.
Maybe @srowen knows the reason. He also changed these files in #23069.

**Comment by dongjoon-hyun** [comment]
Thank you for the confirmation and the pointers!

**Comment by wangyum** [comment]
retest this please

**Comment by juliuszsompolski** [comment]
For readability, would it make sens to split this PR into 2: one to move change the v2.3.4 code to v2.3.5, and the other to start using it?

**Comment by wangyum** [comment]
retest this please

**Comment by juliuszsompolski** [comment]
LGTM, though I would do the change of actually upgrading v2.3.4 to v2.3.5 in a separate PR
cc @gatorsmile

**Comment by gatorsmile** [comment]
retest this please

**Comment by gatorsmile** [comment]
@juliuszsompolski Actually, there does not exist code changes in this component from Hive 2.3.4 to 2.3.5. :-) See the change history. https://github.com/apache/hive/commits/branch-2.3

**Comment by gatorsmile** [comment]
LGTM 

I will merge the code when the tests pass.

**Comment by gatorsmile** [comment]
After the tests pass, we also need to remove [test-hadoop3.2] tag and then retrigger the test again.

**Comment by wangyum** [comment]
retest this please

**Comment by wangyum** [comment]
retest this please

**Comment by wangyum** [comment]
retest this please

**Comment by wangyum** [comment]
retest this please

**Comment by wangyum** [comment]
All tests passed:
hadoop-3.2: https://github.com/apache/spark/pull/24628#issuecomment-498939301
hadoop-2.7: https://github.com/apache/spark/pull/24628#issuecomment-499007269

**Comment by gatorsmile** [comment]
Thanks! Merged to master.

**Pull request closed without merging**